### PR TITLE
removing unfound ecs service from state file

### PR DESCRIPTION
**Pull request by barrytam20**
Changes proposed in this pull request:

* Removing unfound ecs service from statefile instead of throwing an error when running `terraform refresh` to resolve drift

note about acceptance tests: my corporate aws account is not allowed to perform some actions required to run acceptance tests. below is the output, account id/profile have been scrubbed out

Output from acceptance testing:

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSEcsService*'==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSEcsService* -timeout 120m
=== RUN   TestAccAWSEcsServiceDataSource_basic
--- PASS: TestAccAWSEcsServiceDataSource_basic (52.43s)
=== RUN   TestAccAWSEcsService_withARN
--- PASS: TestAccAWSEcsService_withARN (77.58s)
=== RUN   TestAccAWSEcsService_basicImport
--- PASS: TestAccAWSEcsService_basicImport (44.08s)
=== RUN   TestAccAWSEcsService_withUnnormalizedPlacementStrategy
--- PASS: TestAccAWSEcsService_withUnnormalizedPlacementStrategy (48.04s)
=== RUN   TestAccAWSEcsService_withFamilyAndRevision
--- PASS: TestAccAWSEcsService_withFamilyAndRevision (89.89s)
=== RUN   TestAccAWSEcsService_withRenamedCluster
--- PASS: TestAccAWSEcsService_withRenamedCluster (82.52s)
=== RUN   TestAccAWSEcsService_healthCheckGracePeriodSeconds
--- FAIL: TestAccAWSEcsService_healthCheckGracePeriodSeconds (21.41s)
        testing.go:527: Step 2 error: Error applying: 2 errors occurred:
                        * aws_iam_role.ecs_service: 1 error occurred:
                        * aws_iam_role.ecs_service: Error creating IAM Role tf-acc-role-svc-w-hcgps-xvft7vse: AccessDenied: User: arn:aws:sts::aws_account_number:assumed-role/aws_profile is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::aws_account_number:role/tf-acc-role-svc-w-hcgps-xvft7vse
                        status code: 403, request id: 8c8d5c95-c5bf-11e8-9d6d-435ec3fa931d


                        * aws_vpc.main: 1 error occurred:
                        * aws_vpc.main: Error creating VPC: UnauthorizedOperation: You are not authorized to perform this operation.
                        status code: 403, request id: 0dbb3119-7f40-4343-82a6-2a2595c01050




=== RUN   TestAccAWSEcsService_withIamRole
--- FAIL: TestAccAWSEcsService_withIamRole (16.24s)
        testing.go:527: Step 0 error: Error applying: 2 errors occurred:
                        * aws_iam_role.ecs_service: 1 error occurred:
                        * aws_iam_role.ecs_service: Error creating IAM Role tf-acc-role-svc-w-iam-role-4yc2quyc: AccessDenied: User: arn:aws:sts::aws_account_number:assumed-role/aws_profile is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::aws_account_number:role/tf-acc-role-svc-w-iam-role-4yc2quyc
                        status code: 403, request id: 977b8482-c5bf-11e8-82b7-9513ec8e5604


                        * aws_elb.main: 1 error occurred:
                        * aws_elb.main: AccessDenied: User: arn:aws:sts::aws_account_number:assumed-role/aws_profile is not authorized to perform: ec2:CreateSecurityGroup
                        status code: 403, request id: 97a02447-c5bf-11e8-ad34-a9caa0ed026f




=== RUN   TestAccAWSEcsService_withDeploymentValues
--- PASS: TestAccAWSEcsService_withDeploymentValues (51.68s)
=== RUN   TestAccAWSEcsService_withLbChanges
--- FAIL: TestAccAWSEcsService_withLbChanges (67.02s)
        testing.go:527: Step 0 error: Error applying: 2 errors occurred:
                        * aws_iam_role.ecs_service: 1 error occurred:
                        * aws_iam_role.ecs_service: Error creating IAM Role tf-acc-role-svc-w-lbc-l9ziw4xb: AccessDenied: User: arn:aws:sts::aws_account_number:assumed-role/aws_profile is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::aws_account_number:role/tf-acc-role-svc-w-lbc-l9ziw4xb
                        status code: 403, request id: c1862791-c5bf-11e8-8720-f72ce63c99cf


                        * aws_elb.main: 1 error occurred:
                        * aws_elb.main: AccessDenied: User: arn:aws:sts::aws_account_number:assumed-role/aws_profile is not authorized to perform: ec2:CreateSecurityGroup
                        status code: 403, request id: d5ab9fec-c5bf-11e8-afec-f737cfce2486




=== RUN   TestAccAWSEcsService_withEcsClusterName
--- PASS: TestAccAWSEcsService_withEcsClusterName (40.68s)
=== RUN   TestAccAWSEcsService_withAlb
--- FAIL: TestAccAWSEcsService_withAlb (29.47s)
        testing.go:527: Step 0 error: Error applying: 2 errors occurred:
                        * aws_iam_role.ecs_service: 1 error occurred:
                        * aws_iam_role.ecs_service: Error creating IAM Role tf-acc-role-svc-w-alb-jarde82x: AccessDenied: User: arn:aws:sts::aws_account_number:assumed-role/aws_profile is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::aws_account_number:role/tf-acc-role-svc-w-alb-jarde82x
                        status code: 403, request id: 06df72a2-c5c0-11e8-9d6d-435ec3fa931d


                        * aws_vpc.main: 1 error occurred:
                        * aws_vpc.main: Error creating VPC: UnauthorizedOperation: You are not authorized to perform this operation.
                        status code: 403, request id: 186713db-f3c7-4b8b-a738-a3957b61d227




=== RUN   TestAccAWSEcsService_withPlacementStrategy
--- PASS: TestAccAWSEcsService_withPlacementStrategy (134.49s)
=== RUN   TestAccAWSEcsService_withPlacementConstraints
--- PASS: TestAccAWSEcsService_withPlacementConstraints (51.19s)
=== RUN   TestAccAWSEcsService_withPlacementConstraints_emptyExpression
--- PASS: TestAccAWSEcsService_withPlacementConstraints_emptyExpression (39.85s)
=== RUN   TestAccAWSEcsService_withLaunchTypeFargate
--- FAIL: TestAccAWSEcsService_withLaunchTypeFargate (56.44s)
        testing.go:527: Step 0 error: Error applying: 1 error occurred:
                        * aws_vpc.main: 1 error occurred:
                        * aws_vpc.main: Error creating VPC: UnauthorizedOperation: You are not authorized to perform this operation.
                        status code: 403, request id: 845c93fc-0f3e-49fc-8d23-4d2921157e2d




        testing.go:527: Step 0 error: Error applying: 1 error occurred:
                        * aws_vpc.main: 1 error occurred:
                        * aws_vpc.main: Error creating VPC: UnauthorizedOperation: You are not authorized to perform this operation.
                        status code: 403, request id: 0b9abc5b-3a8a-45b8-8a68-fc88590c4285




        testing.go:527: Step 0 error: Error applying: 1 error occurred:
                        * aws_vpc.main: 1 error occurred:
                        * aws_vpc.main: Error creating VPC: UnauthorizedOperation: You are not authorized to perform this operation.
                        status code: 403, request id: 01bf5345-9953-447f-af31-dd2e7c4b67aa




=== RUN   TestAccAWSEcsService_withLaunchTypeEC2AndNetworkConfiguration
--- FAIL: TestAccAWSEcsService_withLaunchTypeEC2AndNetworkConfiguration (56.57s)
        testing.go:527: Step 0 error: Error applying: 1 error occurred:
                        * aws_vpc.main: 1 error occurred:
                        * aws_vpc.main: Error creating VPC: UnauthorizedOperation: You are not authorized to perform this operation.
                        status code: 403, request id: 98fa3fdd-bcef-4080-8d05-3f3917811d28




=== RUN   TestAccAWSEcsService_withDaemonSchedulingStrategy
--- PASS: TestAccAWSEcsService_withDaemonSchedulingStrategy (49.64s)
=== RUN   TestAccAWSEcsService_withReplicaSchedulingStrategy
--- PASS: TestAccAWSEcsService_withReplicaSchedulingStrategy (52.12s)
=== RUN   TestAccAWSEcsService_withServiceRegistries
--- FAIL: TestAccAWSEcsService_withServiceRegistries (21.18s)
        testing.go:527: Step 0 error: Error applying: 1 error occurred:
                        * aws_vpc.test: 1 error occurred:
                        * aws_vpc.test: Error creating VPC: UnauthorizedOperation: You are not authorized to perform this operation.
                        status code: 403, request id: 1799a702-d690-4b34-b2d1-42fe65182355




=== RUN   TestAccAWSEcsService_withServiceRegistries_container
--- FAIL: TestAccAWSEcsService_withServiceRegistries_container (21.69s)
        testing.go:527: Step 0 error: Error applying: 1 error occurred:
                        * aws_vpc.test: 1 error occurred:
                        * aws_vpc.test: Error creating VPC: UnauthorizedOperation: You are not authorized to perform this operation.
                        status code: 403, request id: 4704308a-daa5-4902-b28d-d0ba9b30074a




FAIL
exit status 1
FAIL    github.com/terraform-providers/terraform-provider-aws/aws       1108.355s
make: *** [testacc] Error 1
```

### CODE REVIEW

**Comment by bflad** [comment]
@barrytam20 do you have the error message and/or debug logs from this occurring?

**Comment by barrytam20** [comment]
@bflad this is the error i was seeing when i tried to do a terraform refresh. the service existed in the statefile, but was deleted through the AWS console

```2018-09-29T22:52:10.917Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: 2018/09/29 22:52:10 [DEBUG] [aws-sdk-go] DEBUG: Request ecs/DescribeServices Details:
2018-09-29T22:52:10.917Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: ---[ REQUEST POST-SIGN ]-----------------------------
2018-09-29T22:52:10.917Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: POST / HTTP/1.1
2018-09-29T22:52:10.917Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: Host: ecs.us-east-1.amazonaws.com
2018-09-29T22:52:10.917Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: User-Agent: aws-sdk-go/1.14.26 (go1.9.2; linux; amd64) APN/1.0 HashiCorp/1.0 Terraform/0.11.8-dev
2018-09-29T22:52:10.917Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: Content-Length: 125
2018-09-29T22:52:10.917Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: X-Amz-Target: AmazonEC2ContainerServiceV20141113.DescribeServices
2018-09-29T22:52:10.917Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: Accept-Encoding: gzip
2018-09-29T22:52:10.917Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: 
2018-09-29T22:52:11.242Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: -----------------------------------------------------
2018-09-29T22:52:11.435Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: 2018/09/29 22:52:11 [DEBUG] [aws-sdk-go] DEBUG: Response ecs/DescribeServices Details:
2018-09-29T22:52:11.435Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: ---[ RESPONSE ]--------------------------------------
2018-09-29T22:52:11.435Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: HTTP/1.1 200 OK
2018-09-29T22:52:11.435Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: Connection: close
2018-09-29T22:52:11.435Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: Content-Length: 122
2018-09-29T22:52:11.435Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: Content-Type: application/x-amz-json-1.1
2018-09-29T22:52:11.435Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: Date: Sat, 29 Sep 2018 22:52:13 GMT
2018-09-29T22:52:11.435Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: X-Amzn-Requestid: 4eba7580-c43a-11e8-9e81-79369c8f1b51
2018-09-29T22:52:11.435Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: 
2018-09-29T22:52:11.435Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: 
2018-09-29T22:52:11.435Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: -----------------------------------------------------
2018-09-29T22:52:11.435Z [DEBUG] plugin.terraform-provider-aws_v1.28.0_x4: 2018/09/29 22:52:11 [DEBUG] [aws-sdk-go] {"failures":[{"arn":"arn:aws:ecs:us-east-1:aws_account_id:service/foo-dev-a","reason":"MISSING"}],"services":[]}
2018/09/29 22:52:11 [ERROR] root.foo-service: eval: *terraform.EvalRefresh, err: aws_ecs_service.ecs-svc: No ECS service found: "arn:aws:ecs:us-east-1:aws_account_id:service/foo-dev-a"
2018/09/29 22:52:11 [ERROR] root.foo-service: eval: *terraform.EvalSequence, err: aws_ecs_service.ecs-svc: No ECS service found: "arn:aws:ecs:us-east-1:aws_account_id:service/foo-dev-a"
2018/09/29 22:52:11 [TRACE] [walkRefresh] Exiting eval tree: module.foo-service.aws_ecs_service.ecs-svc```

**Comment by barrytam20** [comment]
hey @bflad is there anything i can do to help with this PR? I haven't really found any other resources with tests that remove unfound resources from the state file

**Comment by bflad** [comment]
@barrytam20 check out `_disappears` acceptance tests 😄  The gist is to:
* `Config: `: Apply a basic configuration
* `Check: `: Use a `TestCheckFunc` to delete from the API
* Add `ExpectsNonEmptyPlan: true`

See also the testing added in #5967

**Comment by barrytam20** [comment]
pushed up test `TestAccAWSEcsService_disappears` 

```make testacc TEST=./aws TESTARGS='-run=TestAccAWSEcsService_disappears'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSEcsService_disappears -timeout 120m
=== RUN   TestAccAWSEcsService_disappears
--- PASS: TestAccAWSEcsService_disappears (23.10s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       45.025s```

File: aws/resource_aws_ecs_service_test.go [hunk 1]
```diff
@@ -148,13 +148,30 @@ func TestAccAWSEcsService_basicImport(t *testing.T) {
...
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSEcsServiceDestroy,
+		Steps: []resource.TestStep{
 			{
-				ResourceName:      resourceName,
-				ImportStateId:     fmt.Sprintf("%s/nonexistent", clusterName),
-				ImportState:       true,
-				ImportStateVerify: false,
-				ExpectError:       regexp.MustCompile(`No ECS service found`),
```
Comment (bflad) [inline aws/resource_aws_ecs_service_test.go]: We actually want to keep this test as its currently the only way we can verify the condition you were originally seeing. We just needed to grab something out of the new error message from import failing (properly!):

```
--- FAIL: TestAccAWSEcsService_basicImport (43.46s)
    testing.go:520: Step 2, expected error:

        1 error occurred:
        	* aws_ecs_service.jenkins (import id: tf-acc-cluster-svc-ye9z6h7j/nonexistent): 1 error occurred:
        	* import aws_ecs_service.jenkins result: nonexistent: import aws_ecs_service.jenkins (id: nonexistent): Terraform detected a resource with this ID doesn't
        exist. Please verify the ID is correct. You cannot import non-existent
        resources using Terraform import.





        To match:

        No ECS service found
```

I went with:

```
ExpectError:       regexp.MustCompile(`Please verify the ID is correct`),
```

File: aws/resource_aws_ecs_service_test.go [hunk 2]
```diff
@@ -803,6 +820,22 @@ func testAccCheckAWSEcsServiceExists(name string, service *ecs.Service) resource
...
+	return func(s *terraform.State) error {
+		conn := testAccProvider.Meta().(*AWSClient).ecsconn
+
+		input := &ecs.DeleteServiceInput{
+			Cluster: aws.String(*service.ClusterArn),
+			Service: aws.String(*service.ServiceName),
+			Force:   aws.Bool(true),
+		}
+
+		_, err := conn.DeleteService(input)
```
Comment (bflad) [inline aws/resource_aws_ecs_service_test.go]: It turns out the ECS service needs some additional checking to ensure its gone `INACTIVE` looking at its Delete function:

```go
func testAccCheckAWSEcsServiceDisappears(service *ecs.Service) resource.TestCheckFunc {
	return func(s *terraform.State) error {
		conn := testAccProvider.Meta().(*AWSClient).ecsconn

		input := &ecs.DeleteServiceInput{
			Cluster: service.ClusterArn,
			Service: service.ServiceName,
			Force:   aws.Bool(true),
		}

		_, err := conn.DeleteService(input)

		if err != nil {
			return err
		}

		// Wait until it's deleted
		wait := resource.StateChangeConf{
			Pending:    []string{"ACTIVE", "DRAINING"},
			Target:     []string{"INACTIVE"},
			Timeout:    10 * time.Minute,
			MinTimeout: 1 * time.Second,
			Refresh: func() (interface{}, string, error) {
				resp, err := conn.DescribeServices(&ecs.DescribeServicesInput{
					Cluster:  service.ClusterArn,
					Services: []*string{service.ServiceName},
				})
				if err != nil {
					return resp, "FAILED", err
				}

				return resp, aws.StringValue(resp.Services[0].Status), nil
			},
		}

		_, err = wait.WaitForState()

		return err
	}
}
```

**Comment by bflad** [comment]
This has been released in [version 1.40.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1400-october-10-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**